staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -477,12 +477,25 @@ func (w *watchCache) WaitUntilFreshAndList(resourceVersion uint64, matchValues [
 	// This isn't the place where we do "final filtering" - only some "prefiltering" is happening here. So the only
 	// requirement here is to NOT miss anything that should be returned. We can return as many non-matching items as we
 	// want - they will be filtered out later. The fact that we return less things is only further performance improvement.
-	// TODO: if multiple indexes match, return the one with the fewest items, so as to do as much filtering as possible.
-	for _, matchValue := range matchValues {
-		if result, err := w.store.ByIndex(matchValue.IndexName, matchValue.Value); err == nil {
-			return result, w.resourceVersion, matchValue.IndexName, nil
+	// If multiple indexes match, return the one with the fewest items, so as to do as much filtering as possible.
+	minIndex := -1
+	minIndexKeysNum := math.MaxUint32
+	for i, matchValue := range matchValues {
+		result, err := w.store.IndexKeys(matchValue.IndexName, matchValue.Value)
```

Comment:
need a profile or something to be sure that this is actually an improvement.

Reply:
Like does our code ever issue requests with multiple indexes possibly being hit?

Reply:
cc @wojtek-t

Reply:
I think I'm against this change.

Not that it's incorrect on its own, but we don't really have multiple indexes and so introducing additional complexity to solve the case which doesn't exist in practice isn't worth it.

We should probably just remove the TODO without doing any changes...